### DISCUSSION TITLE
Installing third party package and use of TMC stacks

### DISCUSSION THREAD

**Discussion by tlaw009** [hsr-project/general#4] [Q&A]
The following content is from an email discussion with Yosuke Matsusaka 

=======================================================

My name is Shengxin (Tony) Luo from the RoboCanes team qualified to participate in this Summer’s RoboCup@Home simulation league. We would appreciate it if you could help us with a couple of questions regarding the logistics of the scoring server. 
 
As I understand it, only code in the registered repository will be considered for scoring. However, our software relies on third party packages both compiled natively from source such as darknet and privately managed such as tmc-melodic-desktop-full. Those packages are essential parts of our software. We would like to know if there is a way, if any, to incorporate those third party packages with the scoring server so that our software would work properly in the competition.

=======================================================

Updates since last discussed: From the TMC stacks, our team mostly uses the its omni_base and whole_body module. Our lab director learned from credible sources that these modules have already been implemented. If any knowledgable party could provide documentation on how to use said module, we would greatly appreciate it.

**Comment by yosuke** [comment]
Hi,
Sorry for the delay.

You can install any software (including non-ROS packages) by writing Dockerfile in the repository.
Your algorithm will run on the scoring server which utilizes nvidia-docker2, so you can also use GPU acceleration using CUDA as well.

This file will help you to understand how to install CUDA in your Dockerfile:
https://github.com/hsr-project/hsrb_robocup_dspl_docker/blob/main/Dockerfile.nvidia

Writing Dockerfile is just as same as writing each Linux command suffixed with the `RUN` statement.

For TMC stacks, if you only use omni_base and whole_body module. The replacement module is already there.

For omni_base, this simulator uses ros_navigation:
```python
navclient = actionlib.SimpleActionClient('/move_base', MoveBaseAction)
```

For whole_body, this simulator uses moveit_commander:
```python
whole_body = moveit_commander.MoveGroupCommander("whole_body_light")
```

I think you can understand how to use it by reading this code:
https://github.com/devrt/robocup-at-home-2021-challenge/blob/main/robocup_challenge/scripts/utils.py

**Reply by tlaw009** [reply]
Thank you, this helped a lot.

Our motion planning pipeline uses move_end_effector_by line()  from whole_body a lot which also comes with the linear and angular weight attribute. This function and the corresponding attributes are not by default available in MoveGroupCommander class, is it implemented as an extension in "whole_body_light" move group? If not, suggestions on how to approach implementing move_end_effector_by_line by ourselves would be much appreciated.

Also, the Dockerfile you were referring to is the one inside the wrs2020_robocup_challenge repo, correct?

**Reply by yosuke** [reply]
The `move_end_effector_by_line` should be replaced by `compute_cartesian_path`:

http://docs.ros.org/en/indigo/api/pr2_moveit_tutorials/html/planning/scripts/doc/move_group_python_interface_tutorial.html#cartesian-paths

For Dockerfile. Yes, you have to place a Dockerfile at top of your challenge repository:
https://github.com/devrt/robocup-at-home-2021-challenge/blob/main/Dockerfile
